### ISSUE TITLE
Default labeling for X axis in timeseries is sparse.

### ISSUE THREAD

**Issue by zaddok** [gonum/plot#296]
When you render a timeseries, no matter how wide you render it, there are only three labels (dates) drawn on the x axis. Probably should include more by default.  (Is there any way to change this programatically?)

![deadlift](https://cloud.githubusercontent.com/assets/20467896/16860763/529b9a44-4a7e-11e6-8584-1bdd4fb67322.png)

**Comment by btracey** [comment]
I agree with the general issue -- the algorithm for selecting tick labels could be better. For real numbers it could do a much better job of trying to find powers of 10 and other round numbers.

**Comment by kortschak** [comment]
@zaddok the distribution of ticks is not dependent on the spatial arrangement of the axis, but rather only the numerical span of the axis. You can see this [here](https://github.com/gonum/plot/blob/master/axis.go#L355). You can see [here](https://play.golang.org/p/n5Xwwx6emu) how changing the start and end dates impact on the tick locations.

Remember that DefaultTicks is just that, a default. Maybe we should have a better heuristic with more knobs, but I think that we will probably still keep an extremely simple default as exists now.

**Comment by zaddok** [comment]
Thanks! I do strongly agree with the general principle of keeping code as simple as possible.  I'll propose my reasons for the idea that, in this case, it may be warranted to improve the default behaviour:

1. If the default behaviour is to create a graph that no-one would ever want, perhaps the default is not adequate.
2. The default behaviour is not friendly for people new to the library, it would be more user friendly if the built in behaviour works out of the box.

Anyway, the tool is awesome either way. You guys have done some great work. This is just my feedback/reaction to the tool as a new user.

**Comment by kortschak** [comment]
Yeah, you are right. The [timeseries](https://github.com/gonum/plot/pull/278) (and indeed [ticks in general](https://github.com/gonum/plot/pull/278)) have PRs for changing behaviour that should make things better.

**Comment by zaddok** [comment]
Aaah great, I'll have a look at the PR's. 

Im still learning the code, my initial reaction however, is that perhaps the `SuggestedTicks` in the default ticker:

    func (DefaultTicks) Ticks(min, max float64) (ticks []Tick) {
       const SuggestedTicks = 3

could be exposed, so that it could be initialised like this:

    p.X.Tick.Marker = plot.UnixTimeTicks{Format: "Jan 2", SuggestedTicks: 5}

**Comment by zaddok** [comment]
Furthermore, I am wondering, Would it be worth submitting a patch for a different ticker options, i.e. "MonthlyTimeTicks", or "WeeklyTimeTicks". (Im going to do it anyway, just not sure if people would be receptive to receiving it into the main code)

**Comment by kortschak** [comment]
They sound reasonable to me.

**Comment by eaburns** [comment]
The issue isn't the number of labels, but label density. Having too many labels in little space makes the plot very noisy (for example look at the y-axis in the plot on the original post here). Having too few makes it hard to comprehend the scale (the x-axis of the same plot).

The underlying problem here is that the ticker code doesn't know the final *size* of the plot. If it did, it would select the number of ticks and the number of labels based on the density. (I think Tufty has recommendations for tick density. We should check.)

The current algorithm is based off an algorithm that I hand-tuned a long time ago when I was interested in 3in x 3in plots for two-column academic papers. I certainly wouldn't say that it is a useless default; it's just not tuned for large plots. That is why I made the tick selection code configurable. More generally, this is why I made lots of things configurable (ticks, vg backbends, plotters, etc.): it's not possible to provide defaults that everyone will like.  

That said, if there is a better default, let's use it.

**Comment by zaddok** [comment]
I'be created my own alternative to the built in unix time series. I'll submit it tomorrow with some sample images to see if people like it. I suspect it'll fill a use case for others as well, if useful then great, if not, no worries.

(And yes, code to measure the widths of the labels would be ideal, but That's no small amount of work for me.)

**Comment by henryx** [comment]
Hi. Any hints for this?

**Comment by Student414** [comment]
```
// MyTicks 自定义Ticks
type MyTicks struct{}

// Ticks returns Ticks in the specified range.
func (MyTicks) Ticks(min, max float64) []plot.Tick {
	if max <= min {
		panic("illegal range")
	}
	var ticks []plot.Tick
	for i := min; i <= max; i++ {
		ticks = append(ticks, plot.Tick{Value: i, Label: strconv.FormatFloat(i, 'f', 0, 64)})
	}
	return ticks
}
.....
p.X.Tick.Marker = MyTicks{}
```

**Comment by cloudufull** [comment]
//添加了一个 Sxlen() 和 Gxlen()   使用前 先自己设置一下就行了
axis.go
```
381 // DefaultTicks is suitable for the Tick.Marker field of an Axis,
382 // it returns a reasonable default set of tick marks.
383 type DefaultTicks struct{}
384 
385 var _ Ticker = DefaultTicks{}
386 
387 var xlen int = 3
388 
389 func (DefaultTicks)Sxlen(l int){
390      xlen = l
391 }
392 
393 func (DefaultTicks)Gxlen() int {
394      return xlen
395 }
396 
397 // Ticks returns Ticks in the specified range.
398 func (DefaultTicks) Ticks(min, max float64) []Tick {
399         if max <= min {
400                 panic("illegal range")
401         }
402 
403         suggestedTicks :=  DefaultTicks.Gxlen(DefaultTicks{})
```
//使用时
 t:=plot.DefaultTicks{}
     t.Sxlen(10)  //在这设置
     xticks := plot.TimeTicks{Format: "01.02\n15:04",Ticker:t}
     p, err := plot.New()
     p.X.Tick.Marker = xticks

**Comment by sbinet** [comment]
FYI, I have also another Ticker implementation, here:
- https://godoc.org/go-hep.org/x/hep/hplot#Ticks

with a plot comparing `gonum/plot` default ticker with "mine":
- https://godoc.org/go-hep.org/x/hep/hplot#example-Ticks
![ticks](https://github.com/go-hep/hep/raw/master/hplot/testdata/ticks_golden.png)
